src/charm.py

```diff
@@ -164,6 +207,53 @@ def __init__(self, *args):
                 self.framework.observe(self.ingress.on.ready, self._on_ingress_ready)
                 self.framework.observe(self.ingress.on.revoked, self._on_ingress_revoked)
 
+    # Frontend TLS handler
+    def _handle_frontend_tls(self, event: EventBase):
+        # Block if the unit is not configured as a frontend service but has the relation
+        if "frontend" not in self.config["services"] and self.model.get_relation(FRONTEND_CERTIFICATES_RELATION_NAME):
+            self.unit.status = BlockedStatus(
+                f"Not a frontend service, please remove {FRONTEND_CERTIFICATES_RELATION_NAME} integration."
+            )
+            return
+
+        # Pre-flight checks
+        if not self.container.can_connect():
+            return
+
+        if not self._relation_created(FRONTEND_CERTIFICATES_RELATION_NAME):
+            return
+
+        # Fetch the assigned certificate and key
+        provider_certificate, private_key = self.certificates.get_assigned_certificate(
+            certificate_request=self._get_certificate_request_attributes()
+        )
+
+        # Set unit to WaitingStatus if certificate or key is not yet available
+        if not provider_certificate or not private_key:
+            logger.info("The certificate is not available yet.")
+            self.unit.status = WaitingStatus("Waiting for certificates to be available")
+            return
+
+        # If either the certificate or key is outdated or missing, update both
+        if self._update_certificates_required(provider_certificate, private_key):
+            self._extra_context.update(FRONTEND_TLS_CONFIGURATION)
+            self._store_certificate(certificate=provider_certificate.certificate)
+            self._store_private_key(private_key=private_key)
+            self._update(event)
+
+    def _on_frontend_certificates_relation_broken(self, event: EventBase):
+        """Handle the frontend-certificates relation broken."""
+        if not self.container.can_connect():
+            event.defer()
```

Comment:
Generally I would discourage the use of defers, they tend to make charms less reliable. I know we don't want to refactor this whole charm here but one way to easily avoid them is through a "central hook" design. 

More specifically, we already have an `update()` central method. There we could look at whether the charm is integrated, has the config option, etc. and make the appropriate decision, like write the certificate or delete it.